data/src/main/java/co/nimblehq/compose/crypto/data/ResponseMapping.kt

```diff
@@ -0,0 +1,41 @@
+package co.nimblehq.compose.crypto.data
+
+import co.nimblehq.compose.crypto.data.model.response.ErrorResponse
+import co.nimblehq.compose.crypto.data.service.JsonApiException
+import co.nimblehq.compose.crypto.data.service.UnknownException
+import co.nimblehq.compose.crypto.data.service.providers.MoshiBuilderProvider
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.transform
+import retrofit2.Response
+
+@OptIn(FlowPreview::class)
+fun <T> Flow<Response<T>>.transform(): Flow<T> = transform { response ->
+    flow {
+        val body = response.body()
+        if (response.isSuccessful && body != null) {
+            emit(body)
+        } else {
+            error(mapError(response))
+        }
+    }
+}
```

Comment:
@hoangnguyen92dn I think we no need to `transform` the value inside `flow` again because we already use it in `repository` 🤔 I have an idea, What about if we create a sealed class `ResponseResult` like `UseCaseResult` to keep the state of value and data, Something like this:
```
sealed class ResponseResult<out T : Any?> {
    class Success<out T : Any>(val data: Any?) : ResponseResult<T>()
    class Error(val exception: Throwable) : ResponseResult<Nothing>()
}
```
```suggestion
fun <T> Response<T>.transform(): ResponseResult<Any> {
    return if (this.isSuccessful && body() != null) {
        ResponseResult.Success(data = body())
    } else {
        ResponseResult.Error(exception = mapError(this))
    }
}
```
So, In the repository, we just emit the value from `ResponseResult` like this 👇 Also, I think by this way It will be easy for testing. What do you think? cc @luongvo 
```
override fun getTrendingList(
        coinIds: String,
        currency: String,
        priceChangePercentage: String,
        itemOrder: String,
        itemPerPage: Int,
        page: Int
    ): Flow<List<CoinItem>> = flow {
        val result = api.getCoins(
            coinIds = coinIds,
            currency = currency,
            priceChangePercentage = priceChangePercentage,
            itemOrder = itemOrder,
            itemPerPage = itemPerPage,
            page = page
        )
        when (result.transform()) {
            is ResponseResult.Success -> emit(result.body()?.toModels().orEmpty())
            is ResponseResult.Error -> error(result)
        }
    }
```

Reply:
@Wadeewee By creating a new Result class, we must always validate the result as Success or Error to return the. We don't handle error on Repository and often throw to UseCase, so it doesn't make sense imo 🤔